src/rapids_singlecell/preprocessing/_neighbors.py

```diff
@@ -356,3 +362,176 @@ def neighbors(
+    random_state
+        A numpy random seed.
+    algorithm
+        The query algorithm to use. Valid options are:
+            * 'brute': Brute-force search that computes distances to all data points, guaranteeing exact results.
+
+            * 'ivfflat': Uses inverted file indexing to partition the dataset into coarse quantizer cells and performs the search within the relevant cells.
+
+            * 'ivfpq': Combines inverted file indexing with product quantization to encode sub-vectors of the dataset, facilitating faster distance computation.
+
+            * 'cagra': Employs the Compressed, Accurate Graph-based search to quickly find nearest neighbors by traversing a graph structure.
+
+        Please ensure that the chosen algorithm is compatible with your dataset and the specific requirements of your search problem.
+    metric
+        A known metric's name or a callable that returns a distance.
+    metric_kwds
+        Options for the metric.
+    key_added
+        If not specified, the neighbors data is stored in .uns['neighbors'],
+        distances and connectivities are stored in .obsp['distances'] and
+        .obsp['connectivities'] respectively.
+        If specified, the neighbors data is added to .uns[key_added],
+        distances are stored in .obsp[key_added+'_distances'] and
+        connectivities in .obsp[key_added+'_connectivities'].
+    copy
+        Return a copy instead of writing to adata.
+
+    Returns
+    -------
+    Depending on `copy`, updates or returns `adata` with the following:
+
+    See `key_added` parameter description for the storage path of
+    connectivities and distances.
+
+        **connectivities** : sparse matrix of dtype `float32`.
+            Weighted adjacency matrix of the neighborhood graph of data
+            points. Weights should be interpreted as connectivities.
+        **distances** : sparse matrix of dtype `float32`.
+            Instead of decaying weights, this stores distances for each pair of
+            neighbors.
+    """
+
+    adata = adata.copy() if copy else adata
+    if adata.is_view:
+        adata._init_as_actual(adata.copy())
+    X = _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
+
+    X_contiguous = _check_neighbors_X(X, algorithm)
+    _check_metrics(algorithm, metric)
+
+    n_neighbors = neighbors_within_batch
+
+    if batch_key is None:
+        raise ValueError("Please provide a batch key to perform batch-balanced KNN.")
+
+    if batch_key not in adata.obs:
+        raise ValueError(f"Batch key '{batch_key}' not present in `adata.obs`.")
+
+    n_obs = adata.shape[0]
+    batch_list = adata.obs[batch_key].values
+    batches = np.unique(batch_list)
+    total_neighbors = n_neighbors * len(batches)
+    knn_dist = cp.zeros((X.shape[0], total_neighbors), dtype=np.float32)
+    knn_indices = cp.zeros_like(knn_dist).astype(int)
+
+    for to_ind in range(len(batches)):
+        batch_to = batches[to_ind]
```

Comment:
```suggestion
    for to_ind, batch_to in enumerate(batches):
```

I see you use `to_ind` later